src/service/middleware/retry/mod.rs

```diff
@@ -0,0 +1,55 @@
+use futures_util::future;
+use http::{Request, Response};
+use hyper::{Body, Error};
+use tower::retry::Policy;
+
+#[derive(Clone)]
+pub enum RetryConfig {
+    None,
+    Simple(usize),
+}
+
+impl Policy<Request<String>, Response<hyper::Body>, hyper::Error> for RetryConfig {
+    type Future = futures_util::future::Ready<Self>;
+
+    fn retry(
+        &self,
+        _req: &Request<String>,
+        result: Result<&Response<Body>, &Error>,
+    ) -> Option<Self::Future> {
+        match self {
+            RetryConfig::None => None,
+            RetryConfig::Simple(count) => match result {
+                Ok(_) => None,
+                Err(_) => {
+                    if *count > 0 {
+                        Some(future::ready(RetryConfig::Simple(count - 1)))
+                    } else {
+                        None
+                    }
+                }
+            },
+        }
+    }
+
+    fn clone_request(&self, req: &Request<String>) -> Option<Request<String>> {
+        match self {
+            RetryConfig::None => None,
+            _ => {
+                // `Request` can't be cloned
+                let mut new_req = Request::builder()
+                    .uri(req.uri())
+                    .method(req.method())
+                    .version(req.version());
+                for (name, value) in req.headers() {
+                    new_req = new_req.header(name, value);
+                }
+
+                let body = req.body().clone();
+                let new_req = new_req.body(body).expect("failed to build request");
```

Comment:
Personally, I'd prefer libraries _not_ panic and instead return errors and let upstream users decide if they want to panic or not (although in the case, I think [Tower will just not retry if we return None](https://docs.rs/tower/latest/src/tower/retry/future.rs.html#78-88)). But I don't know the philosophy on this at the crate level.

Reply:
The reason it has `expect` there is because it should never actually panic there (we are reconstructing Request<String> by cloning components of &Request<String>, so there shouldn't be any invalid values/headers that would cause panic). I will try to clarify with a comment